DOCS-6225 Add DuckDB storage engine page (experimental) (#580)
mariadb-corporation/mariadb-docs: server/server-usage/storage-engines/duckdb-storage-engine.md

---
description: >-
  DuckDB embedded as a MariaDB storage engine for analytical workloads.
  Experimental — alpha maturity, not in mainline MariaDB releases.
---

# DuckDB Storage Engine

{% hint style="danger" %}
**Experimental — for evaluation only.** The DuckDB storage engine is at **alpha** maturity. It is not yet merged into MariaDB Server `main`, not built into MariaDB releases, and not signed in any official package. Interfaces, configuration variables, behavior, and supported operations are subject to change. **Do not use in production.**
{% endhint %}

> The DuckDB storage engine embeds [DuckDB](https://duckdb.org/) — a columnar, vectorized, in-process analytical database — inside MariaDB Server as a loadable plugin (`ha_duckdb.so`). Tables created with `ENGINE=DuckDB` store data in DuckDB's native columnar format, and queries against them are executed by the DuckDB engine. Cross-engine joins between DuckDB and other engines (e.g. InnoDB) are supported in a single `SELECT`.

Reviewer: ```suggestion
The DuckDB storage engine embeds [DuckDB](https://duckdb.org/) — a columnar, vectorized, in-process analytical database engine — inside MariaDB Server as a loadable plugin (`ha_duckdb.so`). Tables created with `ENGINE=DuckDB` store data in DuckDB's native columnar format, and queries against them are executed by the DuckDB engine. Cross-engine joins between DuckDB and other engines (e.g. InnoDB) are supported in a single `SELECT`. 
```

> The engine is ported from Alibaba's [AliSQL](https://github.com/alibaba/AliSQL) DuckDB integration and adapted to MariaDB's handler API, plugin system, and packaging.

Reviewer: ```suggestion
The engine idea has been borrowed from Alibaba's [AliSQL](https://github.com/alibaba/AliSQL) DuckDB integration. The original code has been heavily refactored to leverage MariaDB's handler API, plugin system. MariaDB Duckdb storage engine uses vanilla upstream DuckDB.
```

## Installing

The engine is **not** included in MariaDB releases. There are two ways to obtain it: prebuilt test packages from MariaDB CI, or building from source.

### Option 1: prebuilt test packages

MariaDB CI builds the engine for several Linux distributions and publishes the artifacts on [ci.mariadb.org](https://ci.mariadb.org/). Each CI run has its own job number and a subdirectory per distribution and architecture, for example `amd64-ubuntu-2404-deb-autobake/` or `aarch64-ubuntu-2404-deb-autobake/`.

{% hint style="warning" %}
These packages are **unsigned** and intended for testing only.
{% endhint %}

The CI job number is **not stable** across rebuilds. For the current job URL, see the blog posts under [See also](#see-also).

Example for Ubuntu 24.04 (`amd64`), with a current job number `<JOB>`:

```sh
# Fetch the .deb files for the job
wget -r -np -nH --cut-dirs=2 -A '*.deb' \
  https://ci.mariadb.org/<JOB>/amd64-ubuntu-2404-deb-autobake/

# Install the server and the DuckDB plugin package
sudo dpkg -i mariadb-server*.deb mariadb-plugin-duckdb*.deb || sudo apt -f install
```

### Option 2: build from source

> The engine integration lives on the `bb-11.4-duckdb` branch of [`MariaDB/server`](https://github.com/MariaDB/server). The engine sources themselves are fetched at configure time from [`MariaDB/duckdb-engine`](https://github.com/MariaDB/duckdb-engine) via CMake `FetchContent`. A helper script handles dependency installation and the build.

Reviewer: ```suggestion
The engine integration lives on the `11.4` branch of [`MariaDB/server`](https://github.com/MariaDB/server). The engine sources themselves are fetched at configure time from [`MariaDB/duckdb-engine`](https://github.com/MariaDB/duckdb-engine) via CMake `FetchContent`. A helper script handles dependency installation and the build.
```

```sh

> git clone --recurse-submodules -b bb-11.4-duckdb \

Reviewer: ```suggestion
git clone --recurse-submodules -b 11.4 \
```

  https://github.com/MariaDB/server.git mariadb-server
cd mariadb-server

# Install build prerequisites (requires root)
./storage/duckdb/duckdb/build.sh -D

> # Build and install

Reviewer: ```suggestion
# Build,  install and start MariaDB server
```

> ./storage/duckdb/duckdb/build.sh

Reviewer: ```suggestion
./storage/duckdb/duckdb/build.sh -S
```

# Optional: build a DEB or RPM instead
./storage/duckdb/duckdb/build.sh -p
```

The path `storage/duckdb/duckdb/` reflects the two-level layout: `storage/duckdb/` is the MariaDB build scaffolding, and `storage/duckdb/duckdb/` is the engine source tree populated at configure time.

On Rocky Linux 8, pass `-R` to use `gcc-toolset-12`:

```sh
./storage/duckdb/duckdb/build.sh -R
```

### Enabling the plugin

The DuckDB plugin is not loaded by default. Add a configuration file to the server's drop-in directory:

- **Debian / Ubuntu:** `/etc/mysql/mariadb.conf.d/duckdb.cnf`
- **Rocky / Fedora / openSUSE:** `/etc/my.cnf.d/duckdb.cnf`

```ini
[mysqld]
plugin-maturity=alpha
plugin-load-add=ha_duckdb.so

> duckdb-memory-limit=8G

Reviewer: ```suggestion
duckdb-memory-limit=8G  // the actual value depends on RAM available in your system
```

Reviewer: Plz disregard this one.

- **No XA `PREPARE`.**
- **UTF-8 only** — non-UTF8 charsets fall back to binary comparison. UCA-based MariaDB collations are approximated using DuckDB's `NOCASE` / `NOACCENT` collations.
- **Some MariaDB functions are not pushdown-compatible** — for example, `GROUP_CONCAT()`, `DATE_FORMAT()`, `JSON_CONTAINS()`, `FOUND_ROWS()`, and `LAST_INSERT_ID()` have no DuckDB equivalent or differ in syntax. Such queries fall back to MariaDB-side evaluation.
- **`ALTER COLUMN DROP DEFAULT`** isn't propagated to the DuckDB catalog.
- **`TIMESTAMP` is stored as `TIMESTAMPTZ`** — keep MariaDB and DuckDB timezone settings consistent to avoid value shifts.
- **Cross-engine scan is single-threaded** — only the DuckDB side of a cross-engine query is parallelized.
- **Cross-engine `INSERT` (DuckDB → InnoDB)** has restrictions; see the engine compatibility matrix.

A maintained compatibility matrix is published in the engine repository at [`docs/mariadb-duckdb-incompatibilities.md`](https://github.com/MariaDB/duckdb-engine/blob/main/docs/mariadb-duckdb-incompatibilities.md).

## See also

- [MDEV-39234](https://jira.mariadb.org/browse/MDEV-39234) — upstream feature ticket.

> - [`MariaDB/duckdb-engine`](https://github.com/MariaDB/duckdb-engine) — canonical engine source repository.

Reviewer: The engine code is already in MariaDB server code.

> - [`bb-11.4-duckdb`](https://github.com/MariaDB/server/tree/bb-11.4-duckdb) branch — the MariaDB Server fork containing the integration scaffolding.

Reviewer: ```suggestion
- [`11.4`](https://github.com/MariaDB/server/tree/bb-11.4-duckdb) branch — the MariaDB Server fork containing the integration scaffolding.
```
